File: pkg/client/crypto.go
package client

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/md5"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/pbkdf2"
)

const (
	gcmHeaderMagic     = "OTSGCM1"
	openSSLHeaderMagic = "Salted__"

	gcmSaltSize     = 16
	gcmNonceSize    = 12
	gcmTagSize      = 16
	gcmKDFIterations = 300000

	pbkdf2Iterations = 300000
	aesKeySize       = 32
	aesIVSize        = 16
)

var (
	ErrInvalidCiphertext       = errors.New("invalid ciphertext format or magic header")
	ErrDecryptionFailed        = errors.New("decryption failed: invalid key or payload")
	ErrGCMAuthenticationFailed = errors.New("authentication tag mismatch: payload has been tampered with")
)

// Encrypt encrypts plaintext using AES-256-GCM with PBKDF2-HMAC-SHA256 (300000 iterations).
// It returns a Base64-encoded string representation of the self-contained OTSGCM1 payload.
func Encrypt(passphrase string, plaintext []byte) ([]byte, error) {
	return EncryptGCM(passphrase, plaintext)
}

// Decrypt decrypts encrypted payload by auto-detecting OTSGCM1 (GCM) vs Salted__ (legacy OpenSSL CBC).
func Decrypt(passphrase string, data []byte) ([]byte, error) {
	if len(data) == 0 {
		return nil, nil
	}

	// Auto-decode Base64 if needed
	raw := data
	if bytes.HasPrefix(data, []byte("T1RTR0NN")) || bytes.HasPrefix(data, []byte("U2FsdGVkX1")) || !bytes.HasPrefix(data, []byte(gcmHeaderMagic)) && !bytes.HasPrefix(data, []byte(openSSLHeaderMagic)) {
		decoded, err := decodeBase64Flexible(string(data))
		if err == nil && len(decoded) > 0 {
			raw = decoded
		}
	}

	if bytes.HasPrefix(raw, []byte(gcmHeaderMagic)) {
		return DecryptGCM(passphrase, raw)
	}

	if bytes.HasPrefix(raw, []byte(openSSLHeaderMagic)) {
		return DecryptOpenSSL(passphrase, raw)
	}

	// Retry DecryptGCM then DecryptOpenSSL as fallbacks if header magic was embedded inside raw input
	if unpadded, err := DecryptGCM(passphrase, data); err == nil {
		return unpadded, nil
	} else if errors.Is(err, ErrGCMAuthenticationFailed) {
		return nil, ErrGCMAuthenticationFailed
	}
	if unpadded, err := DecryptOpenSSL(passphrase, data); err == nil {
		return unpadded, nil
	}

	return nil, ErrDecryptionFailed
}

// EncryptGCM encrypts plaintext using AES-256-GCM and PBKDF2-HMAC-SHA256.
// Output binary format: Header ("OTSGCM1", 7B) || Salt (16B) || Nonce (12B) || Ciphertext + Tag (16B)
func EncryptGCM(passphrase string, plaintext []byte) ([]byte, error) {
	if len(plaintext) == 0 {
		return nil, nil
	}

	salt := make([]byte, gcmSaltSize)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
		return nil, fmt.Errorf("generating salt: %w", err)
	}

	key := pbkdf2.Key([]byte(passphrase), salt, gcmKDFIterations, aesKeySize, sha256.New)
	defer zeroBytes(key)

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("creating AES cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("creating GCM mode: %w", err)
	}

	nonce := make([]byte, gcmNonceSize)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, fmt.Errorf("generating nonce: %w", err)
	}

	// Seal appends ciphertext + 16-byte auth tag
	ciphertext := gcm.Seal(nil, nonce, plaintext, nil)

	// Build self-contained payload: Header (7B) || Salt (16B) || Nonce (12B) || Ciphertext+Tag
	raw := make([]byte, 0, len(gcmHeaderMagic)+gcmSaltSize+gcmNonceSize+len(ciphertext))
	raw = append(raw, []byte(gcmHeaderMagic)...)
	raw = append(raw, salt...)
	raw = append(raw, nonce...)
	raw = append(raw, ciphertext...)

	return []byte(base64.StdEncoding.EncodeToString(raw)), nil
}

// DecryptGCM decrypts payload using AES-256-GCM, verifying authentication tag.
func DecryptGCM(passphrase string, data []byte) ([]byte, error) {
	if len(data) == 0 {
		return nil, nil
	}

	raw := data
	if bytes.HasPrefix(data, []byte("T1RTR0NN")) || !bytes.HasPrefix(data, []byte(gcmHeaderMagic)) {
		decoded, err := decodeBase64Flexible(string(data))
		if err == nil && bytes.HasPrefix(decoded, []byte(gcmHeaderMagic)) {
			raw = decoded
		}
	}

	minLen := len(gcmHeaderMagic) + gcmSaltSize + gcmNonceSize + gcmTagSize
	if len(raw) < minLen || !bytes.HasPrefix(raw, []byte(gcmHeaderMagic)) {
		return nil, ErrInvalidCiphertext
	}

	offset := len(gcmHeaderMagic)
	salt := raw[offset : offset+gcmSaltSize]
	offset += gcmSaltSize

	nonce := raw[offset : offset+gcmNonceSize]
	offset += gcmNonceSize

	ciphertext := raw[offset:]

	key := pbkdf2.Key([]byte(passphrase), salt, gcmKDFIterations, aesKeySize, sha256.New)
	defer zeroBytes(key)

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("creating AES cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("creating GCM mode: %w", err)
	}

	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, ErrGCMAuthenticationFailed
	}

	return plaintext, nil
}

// EncryptOpenSSL AES-256-CBC encrypts plaintext using PBKDF2 (SHA-512, 300000 iterations) in standard OpenSSL format.
func EncryptOpenSSL(passphrase string, plaintext []byte) ([]byte, error) {
	if len(plaintext) == 0 {
		return nil, nil
	}

	salt := make([]byte, 8)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
		return nil, fmt.Errorf("generating salt: %w", err)
	}

	derived := pbkdf2.Key([]byte(passphrase), salt, pbkdf2Iterations, aesKeySize+aesIVSize, sha512.New)
	key := derived[:aesKeySize]
	iv := derived[aesKeySize:]

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("creating cipher: %w", err)
	}

	padded := pkcs7Padding(plaintext, aes.BlockSize)
	ciphertext := make([]byte, len(padded))

	mode := cipher.NewCBCEncrypter(block, iv)
	mode.CryptBlocks(ciphertext, padded)

	header := append([]byte(openSSLHeaderMagic), salt...)
	raw := append(header, ciphertext...)
	return []byte(base64.StdEncoding.EncodeToString(raw)), nil
}

// DecryptOpenSSL AES-256-CBC decrypts OpenSSL formatted ciphertext ("Salted__" + 8-byte salt + data).
func DecryptOpenSSL(passphrase string, data []byte) ([]byte, error) {
	if len(data) == 0 {
		return nil, nil
	}

	raw := data
	if bytes.HasPrefix(data, []byte("U2FsdGVkX1")) || !bytes.HasPrefix(data, []byte(openSSLHeaderMagic)) {
		decoded, err := decodeBase64Flexible(string(data))
		if err == nil && bytes.HasPrefix(decoded, []byte(openSSLHeaderMagic)) {
			raw = decoded
		}
	}

	if len(raw) < 16 || !bytes.HasPrefix(raw, []byte(openSSLHeaderMagic)) {
		return nil, ErrInvalidCiphertext
	}

	salt := raw[8:16]
	ciphertext := raw[16:]

	if len(ciphertext)%aes.BlockSize != 0 {
		return nil, ErrInvalidCiphertext
	}

	// Try PBKDF2 first
	derived := pbkdf2.Key([]byte(passphrase), salt, pbkdf2Iterations, aesKeySize+aesIVSize, sha512.New)
	key := derived[:aesKeySize]
	iv := derived[aesKeySize:]

	if unpadded, err := decryptAESCBC(key, iv, ciphertext); err == nil {
		return unpadded, nil
	}

	// Fallback to legacy OpenSSL EVP_BytesToKey (MD5 digest) for historical payloads
	keyMD5, ivMD5 := openSSLMD5KDF([]byte(passphrase), salt)
	if unpadded, err := decryptAESCBC(keyMD5, ivMD5, ciphertext); err == nil {
		return unpadded, nil
	}

	return nil, ErrDecryptionFailed
}

func decryptAESCBC(key, iv, ciphertext []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	plaintext := make([]byte, len(ciphertext))
	mode := cipher.NewCBCDecrypter(block, iv)
	mode.CryptBlocks(plaintext, ciphertext)

	return pkcs7Unpadding(plaintext, aes.BlockSize)
}

func openSSLMD5KDF(passphrase, salt []byte) (key, iv []byte) {
	var prev []byte
	var derived []byte
	for len(derived) < aesKeySize+aesIVSize {
		h := md5.New()
		h.Write(prev)
		h.Write(passphrase)
		h.Write(salt)
		prev = h.Sum(nil)
		derived = append(derived, prev...)
	}
	return derived[:aesKeySize], derived[aesKeySize : aesKeySize+aesIVSize]
}

func pkcs7Padding(data []byte, blockSize int) []byte {
	padding := blockSize - (len(data) % blockSize)
	padText := bytes.Repeat([]byte{byte(padding)}, padding)
	return append(data, padText...)
}

func pkcs7Unpadding(data []byte, blockSize int) ([]byte, error) {
	length := len(data)
	if length == 0 || length%blockSize != 0 {
		return nil, ErrInvalidCiphertext
	}
	padding := int(data[length-1])
	if padding == 0 || padding > blockSize || padding > length {
		return nil, ErrInvalidCiphertext
	}
	var match byte
	for i := length - padding; i < length; i++ {
		match |= data[i] ^ byte(padding)
	}
	if subtle.ConstantTimeByteEq(match, 0) != 1 {
		return nil, ErrInvalidCiphertext
	}
	return data[:length-padding], nil
}

func decodeBase64Flexible(s string) ([]byte, error) {
	if decoded, err := base64.StdEncoding.DecodeString(s); err == nil {
		return decoded, nil
	}
	if decoded, err := base64.RawURLEncoding.DecodeString(s); err == nil {
		return decoded, nil
	}
	return base64.URLEncoding.DecodeString(s)
}

func zeroBytes(b []byte) {
	for i := range b {
		b[i] = 0
	}
}

File: src/crypto.ts
import base64 from "base64-js";

const gcmBanner = new Uint8Array(new TextEncoder().encode("OTSGCM1"));
const opensslBanner = new Uint8Array(new TextEncoder().encode("Salted__"));

const gcmPbkdf2Params = { hash: "SHA-256", iterations: 300000, name: "PBKDF2" };
const cbcPbkdf2Params = { hash: "SHA-512", iterations: 300000, name: "PBKDF2" };

/**
 * Decrypts ciphertext by auto-detecting OTSGCM1 (GCM) vs Salted__ (legacy OpenSSL CBC).
 */
function dec(cipherText: string, passphrase: string): Promise<string> {
  return decrypt(passphrase, cipherText);
}

/**
 * Encrypts plaintext using modern AES-256-GCM with PBKDF2-HMAC-SHA256.
 */
function enc(plainText: string, passphrase: string): Promise<string> {
  return encryptGCM(passphrase, plainText);
}

function decrypt(passphrase: string, encData: string): Promise<string> {
  let data: Uint8Array;
  try {
    data = base64.toByteArray(encData);
  } catch (e) {
    return Promise.reject(new Error("Invalid base64 payload"));
  }

  // Check OTSGCM1 header (7 bytes)
  if (hasPrefix(data, gcmBanner)) {
    return decryptGCM(passphrase, data);
  }

  // Check Salted__ header (8 bytes)
  if (hasPrefix(data, opensslBanner)) {
    return decryptCBC(passphrase, data);
  }

  // Fallback attempt: try GCM then CBC
  return decryptGCM(passphrase, data).catch(() => decryptCBC(passphrase, data));
}

function encryptGCM(passphrase: string, plainData: string): Promise<string> {
  const salt = window.crypto.getRandomValues(new Uint8Array(16));
  const nonce = window.crypto.getRandomValues(new Uint8Array(12));

  return window.crypto.subtle
    .importKey("raw", new TextEncoder().encode(passphrase), "PBKDF2", false, ["deriveKey"])
    .then((baseKey) =>
      window.crypto.subtle.deriveKey(
        { ...gcmPbkdf2Params, salt },
        baseKey,
        { length: 256, name: "AES-GCM" },
        false,
        ["encrypt"],
      ),
    )
    .then((aesKey) =>
      window.crypto.subtle.encrypt(
        { iv: nonce, name: "AES-GCM" },
        aesKey,
        new TextEncoder().encode(plainData),
      ),
    )
    .then((encData) => {
      const banner = gcmBanner;
      const cipherBytes = new Uint8Array(encData);
      const data = new Uint8Array(banner.byteLength + salt.byteLength + nonce.byteLength + cipherBytes.byteLength);

      let offset = 0;
      data.set(banner, offset); offset += banner.byteLength;
      data.set(salt, offset); offset += salt.byteLength;
      data.set(nonce, offset); offset += nonce.byteLength;
      data.set(cipherBytes, offset);

      return base64.fromByteArray(data);
    });
}

function decryptGCM(passphrase: string, data: Uint8Array): Promise<string> {
  if (data.byteLength < 7 + 16 + 12 + 16) {
    return Promise.reject(new Error("Invalid GCM payload size"));
  }

  const salt = data.slice(7, 23);
  const nonce = data.slice(23, 35);
  const ciphertext = data.slice(35);

  return window.crypto.subtle
    .importKey("raw", new TextEncoder().encode(passphrase), "PBKDF2", false, ["deriveKey"])
    .then((baseKey) =>
      window.crypto.subtle.deriveKey(
        { ...gcmPbkdf2Params, salt },
        baseKey,
        { length: 256, name: "AES-GCM" },
        false,
        ["decrypt"],
      ),
    )
    .then((aesKey) =>
      window.crypto.subtle.decrypt(
        { iv: nonce, name: "AES-GCM" },
        aesKey,
        ciphertext,
      ),
    )
    .then((decrypted) => new TextDecoder("utf8").decode(decrypted));
}

function decryptCBC(passphrase: string, data: Uint8Array): Promise<string> {
  const salt = data.slice(8, 16);
  const ciphertext = data.slice(16);

  return deriveKeyCBC(passphrase, salt)
    .then(({ iv, key }) =>
      window.crypto.subtle.decrypt({ iv, name: "AES-CBC" }, key, ciphertext),
    )
    .then((decrypted) => new TextDecoder("utf8").decode(decrypted));
}

function deriveKeyCBC(passphrase: string, salt: Uint8Array): Promise<{ iv: Uint8Array; key: CryptoKey }> {
  return window.crypto.subtle
    .importKey("raw", new TextEncoder().encode(passphrase), "PBKDF2", false, ["deriveBits"])
    .then((passwordKey) =>
      window.crypto.subtle.deriveBits({ ...cbcPbkdf2Params, salt }, passwordKey, 384),
    )
    .then((key) =>
      window.crypto.subtle
        .importKey("raw", key.slice(0, 32), { name: "AES-CBC" }, false, ["encrypt", "decrypt"])
        .then((aesKey) => ({ iv: key.slice(32, 48), key: aesKey })),
    );
}

function hasPrefix(arr: Uint8Array, prefix: Uint8Array): boolean {
  if (arr.byteLength < prefix.byteLength) return false;
  for (let i = 0; i < prefix.byteLength; i++) {
    if (arr[i] !== prefix[i]) return false;
  }
  return true;
}

export default { dec, enc };
